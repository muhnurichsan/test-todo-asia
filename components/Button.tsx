interface ButtonProps {
  color?: "red" | "yellow" | "grey" | "white";
  label?: string;
  type: "submit" | "button";
  onClick?: () => void;
  fullWidth?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  color,
  label,
  type,
  onClick,
  fullWidth,
}) => {
  return (
    <button
      onClick={onClick}
      type={type}
      className={`
      py-3 
      px-3 
      inline-flex 
      justify-center 
      items-center 
      gap-2 rounded-md
      border 
      font-semibold
    ${
      color === "red"
        ? "text-white bg-red-600 hover:bg-red-500 focus:ring-red-500"
        : color === "grey"
        ? "text-white bg-slate-600 focus:ring-slate-300 hover:bg-slate-500 "
        : color === "yellow"
        ? " text-yellow-500 bg-yellow-100 hover:bg-yellow-100 focus:ring-yellow-500"
        : color === "white"
        ? "bg-transparent text-black"
        : "text-white bg-blue-600 hover:bg-blue-500 focus:ring-blue-500"
    }
      ring-offset-white 
      focus:outline-none 
      focus:ring-2 
      focus:ring-offset-2 
      transition-all 
      ${fullWidth ? "w-full" : "w-fit"}
      `}
    >
      <p className="block text-xs md:text-md">{label}</p>
    </button>
  );
};

export default Button;
