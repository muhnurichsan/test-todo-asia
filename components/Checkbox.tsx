interface CheckboxProps {
  checked?: boolean;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Checkbox: React.FC<CheckboxProps> = ({ checked, onChange }) => {
  return (
    <input
      type="checkbox"
      defaultChecked={checked}
      onChange={onChange}
      className="w-4"
    />
  );
};

export default Checkbox;
