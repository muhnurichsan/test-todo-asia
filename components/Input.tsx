interface InputProps {
  value?: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  type: "text";
  name?: string;
  placeholder?: string;
  onClick?: (e: any) => void;
  onKeyUp?: (e: any) => void;
}

const Input: React.FC<InputProps> = ({
  onChange,
  value,
  type,
  name,
  placeholder,
  onClick,
  onKeyUp,
}) => {
  return (
    <div className="relative w-full border-[1px] rounded-md">
      <input
        onKeyUp={onKeyUp}
        type={type}
        value={value}
        onChange={onChange}
        name={name}
        id={name}
        onClick={onClick}
        className={`p-2 block 
        w-full 
        h-full
        rounded-md text-sm 
        placeholder:text-xs
        focus:border-black
        `}
        placeholder={placeholder}
      />
    </div>
  );
};

export default Input;
