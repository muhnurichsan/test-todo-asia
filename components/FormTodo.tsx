import Button from "./Button";
import Input from "./Input";

interface FormTodoProps {
  onSubmit: (e: React.FormEvent) => void;
  handleInputTask: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleSortTodo: () => void;
}

const FormTodo: React.FC<FormTodoProps> = ({
  onSubmit,
  handleInputTask,
  handleSortTodo,
}) => {
  return (
    <form onSubmit={onSubmit} className="mb-4">
      <h1 className="text-black font-bold text-xl">Todo List</h1>
      <div className="flex mt-4 gap-5">
        <Input
          placeholder="masukkan nama task"
          type="text"
          onChange={handleInputTask}
        ></Input>
        <Button type="submit" label="Add"></Button>
        <div className="w-1/2">
          <Button
            type="button"
            label="Descending Order"
            color="grey"
            onClick={handleSortTodo}
          ></Button>
        </div>
      </div>
    </form>
  );
};

export default FormTodo;
