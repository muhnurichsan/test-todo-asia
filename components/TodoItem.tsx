import { useCallback, useState } from "react";
import Button from "./Button";
import Input from "./Input";
import Checkbox from "./Checkbox";
import axios from "axios";
import useTodos from "@/hooks/useTodos";
import { TodoTypes } from "@/services/data-types";

interface TodoListProps {
  data: TodoTypes;
}

const TodoList: React.FC<TodoListProps> = ({ data }) => {
  const { mutate: mutateTodo } = useTodos();
  const [taskInput, setTaskInput] = useState(data.task);
  const [taskClicked, setTaskClicked] = useState(false);

  const handleUpdateTask = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      e.stopPropagation();

      try {
        await axios.put(`/api/todo/edit/${data.id}`, {
          task: taskInput,
        });
      } catch (error) {
        console.log(error);
      } finally {
        mutateTodo();
      }
    },
    [data.id, taskInput, mutateTodo]
  );

  const handleDoneTodo = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      try {
        await axios.put(`/api/todo/edit/${data.id}`, {
          done: e.target.checked,
        });
        mutateTodo();
      } catch (error) {
        console.log(error);
      }
    },
    [data.id, mutateTodo]
  );

  const handleDeleteTodo = useCallback(async () => {
    try {
      await axios.delete(`/api/todo/delete/${data.id}`);
      mutateTodo();
    } catch (error) {
      console.log(error);
    }
  }, [data.id, mutateTodo]);

  return (
    <li
      onClick={() => {
        setTaskClicked(false);
      }}
      className="flex justify-between items-center"
    >
      <div className="flex gap-3">
        <Checkbox checked={data.done} onChange={handleDoneTodo}></Checkbox>
        {taskClicked ? (
          <Input
            type="text"
            onChange={(e) => {
              setTaskInput(e.target.value);
            }}
            value={taskInput}
            onKeyUp={handleUpdateTask}
            onClick={(e) => {
              e.stopPropagation();
            }}
          ></Input>
        ) : (
          <p
            className={`text-md font-semibold ${data.done && "line-through"}`}
            onClick={(e: any) => {
              e.stopPropagation();
              setTaskClicked(true);
            }}
          >
            {data.task}
          </p>
        )}
      </div>
      <div>
        <Button
          type="button"
          color="red"
          label="Delete"
          onClick={handleDeleteTodo}
        ></Button>
      </div>
    </li>
  );
};

export default TodoList;
