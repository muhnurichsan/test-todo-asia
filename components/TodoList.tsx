import { useState } from "react";
import Button from "./Button";
import Input from "./Input";
import Checkbox from "./Checkbox";

interface TodoListProps {
  data: Record<string, any>;
}

const TodoList: React.FC<TodoListProps> = ({ data }) => {
  const [titleClicked, setTitleClicked] = useState(false);
  const [checked, setChecked] = useState(false);

  return (
    <li
      onClick={() => {
        setTitleClicked(false);
      }}
      className="flex justify-between items-center"
    >
      <div className="flex gap-3">
        <Checkbox
          checked={data.done}
          onChange={() => {
            setChecked(!checked);
          }}
        ></Checkbox>
        {titleClicked ? (
          <Input
            label="task"
            type="text"
            value={data.task}
            onChange={() => {}}
            onClick={(e) => {
              e.stopPropagation();
            }}
          ></Input>
        ) : (
          <p
            className={`text-md font-semibold ${data.done && "line-through"}`}
            onClick={(e: any) => {
              e.stopPropagation();
              setTitleClicked(true);
            }}
          >
            {data.task}
          </p>
        )}
      </div>
      <div>
        <Button type="button" color="red" label="Delete"></Button>
      </div>
    </li>
  );
};

export default TodoList;
