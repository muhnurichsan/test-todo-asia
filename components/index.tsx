import Button from "./Button";
import Checkbox from "./Checkbox";
import FormTodo from "./FormTodo";
import Input from "./Input";
import TodoItem from "./TodoItem";

export { Button, Checkbox, FormTodo, Input, TodoItem };
