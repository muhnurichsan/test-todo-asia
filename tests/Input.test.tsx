import renderer from "react-test-renderer";
import { Input } from "../components";

test("should render component and match snapshot", () => {
  const component = renderer.create(<Input type="text" onChange={() => {}} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
