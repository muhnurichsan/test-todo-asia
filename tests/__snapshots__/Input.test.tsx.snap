// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render component and match snapshot 1`] = `
<div
  className="relative w-full border-[1px] rounded-md"
>
  <input
    className="p-2 block 
        w-full 
        h-full
        rounded-md text-sm 
        placeholder:text-xs
        focus:border-black
        "
    onChange={[Function]}
    type="text"
  />
</div>
`;
